### PR TITLE
[tests] Add a cecil-test to verify capitalization.

### PR DESCRIPTION
**Pull request by rolfbjarne**
Add a cecil test to verify a few common capitalization issues:

* All APIs must start with an upper-cased letter (except parameter names, which must start with a lower-cased letter).
* The term "URL" should always be capitalized as "Url".

### CODE REVIEW

File: tests/cecil-tests/Helper.cs [hunk 1]
```diff
@@ -219,6 +224,139 @@ public static IEnumerable<ICustomAttributeProvider> EnumerateAttributeProviders
...
+				return true;
+			var setter = property.SetMethod;
+			if (setter is not null && IsPubliclyVisible (setter))
+				return true;
+			return false;
+		}
+
+		public static IEnumerable<MemberReference> EnumeratePublicMembers (this AssemblyDefinition assembly, Func<MemberReference, bool>? filter = null)
+		{
+			foreach (var item in assembly.EnumerateTypes (filter)) {
```
Comment (mandel-macaque) [inline tests/cecil-tests/Helper.cs:329]: I think we could make things cleaner if we did a function composition with the filter and the IsPubliclyVisible functions to reduce the need to iterate more than needed. Both are Func<MemeberReference, bool> and we could combine them like this:

```csharp
public static Func<T, bool> Compose<T>(this Func<T, bool>? func1, Func<T, bool> func2)
{
	return func1 is null ? func2 : mr => func1 (mr) && func2 (mr);
}
```

Once we have the ability to compose the filter function with a similar one, we can create a IsVisible function like the following one:

```csharp
public static bool IsVisible (MemberReference memberReference) => memberReference switch {
		PropertyDefinition p => p.IsVisible (),
		MethodDefinition m => m.IsVisible (),
		_ => throw new ArgumentException ()
};
```
Taking those two functions + the Concate from Linq, this method could be:
```
public static IEnumerable<MemberReference> EnumeratePublicMembers (this AssemblyDefinition assembly, Func<MemberReference, bool>? filter = null)
{
	var compose = filter.Compose (IsVisible);
	return assembly.EnumerateProperties (compose)
		.Concat<MemberReference> (assembly.EnumerateMethods (compose));
}
```

File: tests/cecil-tests/Helper.cs [hunk 2]
```diff
@@ -219,6 +224,139 @@ public static IEnumerable<ICustomAttributeProvider> EnumerateAttributeProviders
...
+				yield return item;
+
+			foreach (var item in assembly.EnumerateMethods (filter))
+				yield return item;
+
+			foreach (var item in assembly.EnumerateProperties (filter))
+				yield return item;
+
+			foreach (var item in assembly.EnumerateEvents (filter))
+				yield return item;
```
Comment (mandel-macaque) [inline tests/cecil-tests/Helper.cs:242]: You can use Concat from Linq here.

**Review by mandel-macaque — APPROVED** [review]
Approving with a small comment on how we could write this in a more functional approach to reduce all those loops (and some extra iterations).

File: tests/cecil-tests/Helper.cs (hunk 2 above)
Comment (rolfbjarne) [reply tests/cecil-tests/Helper.cs:242]: That doesn't really line up all that nicely...

File: tests/cecil-tests/Helper.cs (hunk 1 above)
Comment (rolfbjarne) [reply tests/cecil-tests/Helper.cs]: OK, simplified it somewhat.

**Review by dalexsoto — APPROVED** [review]

File: tests/cecil-tests/ApiTest.cs [hunk 3]
```diff
@@ -49,5 +49,116 @@ static void AddFailure (ref List<string>? failures, string failure)
...
+							continue;
+					}
+
+					if (!char.IsUpper (member.Name [0]) && !IsAcceptableCapitalization (member)) {
+						var msg = $"The {member.GetType ().Name.Replace ("Definition", "").ToLower ()} '{member.FullName}' has incorrect capitalization: first letter is not upper case.";
+						failures [msg] = new (msg, member.RenderLocation ());
+					}
+				}
+
+				foreach (var method in assembly.EnumerateMethods (Helper.IsPubliclyVisible)) {
```
Comment (stephen-hawley) [inline tests/cecil-tests/ApiTest.cs:94]: It looks like `EnumeratePublicMembers` already hands you methods. Maybe you want to move the foreach on the parameters up above and add an `else if (member is MethodDefinition method) { }` around it.


Also, will this work in inner classes?

**Review by stephen-hawley — CHANGES_REQUESTED** [review]

File: tests/cecil-tests/ApiTest.cs (hunk 3 above)
Comment (rolfbjarne) [reply tests/cecil-tests/ApiTest.cs]: @stephen-hawley should be fixed now

**Review by stephen-hawley — APPROVED** [review]

**Pull request merged**